### PR TITLE
Initial attempt at describing the event redaction algorithm

### PR DESCRIPTION
**Pull request by leonerd**
@NegativeMjark - can you please review this for technical accuracy.

@Kegsay - once it contains no lies, can you review it for wording style.

Thanks.

### CODE REVIEW

**Comment by NegativeMjark** [comment]
This looks like an accurate translation of https://github.com/matrix-org/synapse/blob/master/synapse/events/utils.py

I'll leave it to @Kegsay to check the documentation style.

File: specification/event_signing.rst [hunk 1]
```diff
@@ -190,9 +190,63 @@ in the event JSON in a ``hash`` object under a ``sha256`` key.
             event_json_object["unsigned"] = unsigned
         return event_json_object
 
-Then all non-essential keys are stripped from the event object, and the
-resulting object which included the ``hash`` key is signed using the JSON
-signing algorithm
+The event is then stripped of all non-essential keys both at the top level and
+within the ``content`` object. Essential top-level keys given below; any
+top-level keys not in this list are removed.
```
Comment (Kegsay) [inline specification/event_signing.rst]: > Essential top-level keys given below; any top-level keys not in this list are removed.

Any top-level keys not in the following list MUST be removed:

File: specification/event_signing.rst [hunk 2]
```diff
@@ -190,9 +190,63 @@ in the event JSON in a ``hash`` object under a ``sha256`` key.
...
+
+.. code::
+
+    auth_events
+    depth
```
Comment (Kegsay) [inline specification/event_signing.rst]: Is `depth` really a protocol thing? I recall seeing other areas which merely "suggest" implementations keep a `depth`.

File: specification/event_signing.rst [hunk 3]
```diff
@@ -190,9 +190,63 @@ in the event JSON in a ``hash`` object under a ``sha256`` key.
...
+    prev_state
+    room_id
+    sender
+    signatures
+    state_key
+    type
+
+A new ``content`` object is constructed for the resulting event that contains
+only the essential keys of the original event. If the original event lacked a
+``content`` object at all, a new blank one is created for it.
```
Comment (Kegsay) [inline specification/event_signing.rst]: Be more specific when you say "blank one". I presume you mean "an empty JSON object" but bear in mind our beloved protocol has 0-length strings, etc so confusion is a real possibility.

File: specification/event_signing.rst [hunk 4]
```diff
@@ -190,9 +190,63 @@ in the event JSON in a ``hash`` object under a ``sha256`` key.
...
+      events
+      events_default
+      kick
+      redact
+      state_default
+      users
+      users_default
+
+The resulting stripped object with the new ``content`` object and the original
+``hashes`` key is then signed using the JSON signing algorithm.
```
Comment (Kegsay) [inline specification/event_signing.rst]: > is then signed using the JSON signing algorithm

is then signed using the JSON signing algorithm outlined below:

File: specification/event_signing.rst [hunk 5]
```diff
@@ -190,9 +190,63 @@ in the event JSON in a ``hash`` object under a ``sha256`` key.
...
+    prev_events
+    prev_state
+    room_id
+    sender
+    signatures
+    state_key
+    type
+
+A new ``content`` object is constructed for the resulting event that contains
+only the essential keys of the original event. If the original event lacked a
```
Comment (Kegsay) [inline specification/event_signing.rst]: > only the essential keys of the original event

only the essential keys from the original ``content``.

File: specification/event_signing.rst (hunk 1 above)
Comment (leonerd) [inline specification/event_signing.rst]: Done

File: specification/event_signing.rst (hunk 2 above)
Comment (leonerd) [inline specification/event_signing.rst]: I'm not sure if the key is required for all events; it's simply that if such a key exists then redaction will leave it undisturbed and not remove it.

File: specification/event_signing.rst (hunk 3 above)
Comment (leonerd) [inline specification/event_signing.rst]: Changed.

File: specification/event_signing.rst (hunk 4 above)
Comment (leonerd) [inline specification/event_signing.rst]: Done

File: specification/event_signing.rst (hunk 5 above)
Comment (leonerd) [inline specification/event_signing.rst]: Done.

**Comment by leonerd** [comment]
@kegan PTAL

File: specification/event_signing.rst [hunk 6]
```diff
@@ -190,9 +190,64 @@ in the event JSON in a ``hash`` object under a ``sha256`` key.
             event_json_object["unsigned"] = unsigned
         return event_json_object
 
-Then all non-essential keys are stripped from the event object, and the
-resulting object which included the ``hash`` key is signed using the JSON
-signing algorithm
+The event is then stripped of all non-essential keys both at the top level and
+within the ``content`` object. Any top-level keys not in he following list MUST
```
Comment (Kegsay) [inline specification/event_signing.rst]: Typo. s/he/the/

**Comment by Kegsay** [comment]
LGTM other than typo.

**Pull request merged**